### PR TITLE
Service type runit

### PR DESCRIPTION
**Pull request by justizin**
Implement runit service type, following the lead of SimpleFinance/chef-zookeeper

### CODE REVIEW

**Comment by mathyourlife** [comment]
test converges on both ubuntu 14.04 and debian 7.8 fail to start kafka

``` diff
diff --git a/.kitchen.yml b/.kitchen.yml
index 52c0417..8db8fcf 100644
--- a/.kitchen.yml
+++ b/.kitchen.yml
@@ -16,7 +16,7 @@ platforms:
       zookeeper:
         service_style: "upstart"
       apache_kafka:
-        service_style: "upstart"
+        service_style: "runit"
   - name: debian-7.8
     driver_config:
       box: 'chef/debian-7.8'
@@ -29,7 +29,7 @@ platforms:
       zookeeper:
         service_style: "runit"
       apache_kafka:
-        service_style: "init.d"
+        service_style: "runit"

 suites:
   - name: default
```

I'm Using ChefDK to run the tests

```
kitchen test
```

**Comment by justizin** [comment]
Good catch!

I'm going to take a look at what the zk cookbook does with this, because runit doesn't really give you a success or failure to start - a test that tries to run the initscript may fail because runit has already started kafka, and an appropriate fix may be to conditionally disable that script and check for a sign that kafka is truly running.

**Comment by mathyourlife** [comment]
I haven't had a chance to dig into it other than running the tests.  Is it possible the source /etc/default/kafka is required in the runit sh script?

**Comment by justizin** [comment]
Shouldn't be, what I'm saying is that there is no external difference between starting a runit service successfully and starting it unsuccessfully, if the script fails, it will loop forever - the shell command to signal starting will succeed unless the user lacks permission.

What ServerSpec seems to be checking is if 'service kafka status' will return info about a running service, which should fail.

I'll dig back into this tomorrow, my primary way of testing the greater goal is to vagrant destroy && vagrant up, I thought this was working when I submitted the PR, but it needs improvement in the tests at a minimum.

Thanks for taking a look!

**Comment by mathyourlife** [comment]
Sure thing. Thanks for the PR

**Comment by justizin** [comment]
Interesting note, I have a merge branch in my fork with both of my PR branches, and 'kitchen test' just passed on that branch.  running it specifically on this one.

I did have to drop the box url in .kitchen.yml because it's pointing at a box that no longer exists, but 'ubuntu/trusty64' on its' own is able to pull an upstream box nowadays.  I wonder if this also impacts your result.

Will keep working at it. :)

**Comment by justizin** [comment]
Current output on my machine, at least for the debian test, I seem to have lost ubuntu in my buffer and will tee the output on next run:

```
-----> serverspec installed (version 2.18.0)
       /opt/chef/embedded/bin/ruby -I/tmp/verifier/suites/serverspec -I/tmp/verifier/gems/gems/rspec-support-3.3.0/lib:/tmp/verifier/gems/gems/rspec-core-3.3.0/lib /opt/chef/embedded/bin/rspec --pattern /tmp/verifier/suites/serverspec/\*\*/\*_spec.rb --color --format documentation --default-path /tmp/verifier/suites/serverspec

       Kafka broker
         is listening on port 9092
         is listening on port 9093
         has a running service of kafka

       Finished in 0.06483 seconds (files took 0.18994 seconds to load)
       3 examples, 0 failures

       Finished verifying <default-debian-78> (0m29.80s).
-----> Destroying <default-debian-78>...
       ==> default: Forcing shutdown of VM...
       ==> default: Destroying VM and associated drives...
       Vagrant instance <default-debian-78> destroyed.
       Finished destroying <default-debian-78> (0m3.25s).
       Finished testing <default-debian-78> (23m3.93s).
-----> Kitchen is finished. (39m53.18s)
[jryan@paradox apache_kafka]$ git branch
  jryan_work
  master
  paramaterize_script_list
* service_type_runit
```

**Comment by mathyourlife** [comment]
interesting... And your `service_style` is set to "runit" in .kitchen.yml?

**Comment by justizin** [comment]
Yep!

I also had to remove the urls to the vagrant box images, as they're unavailable - I suspect your local testing environment has a copy of them from the past.  Both operating systems can be fetched by name from the main repositories.

``` diff
diff --git a/.kitchen.yml b/.kitchen.yml
index 52c0417..4793bed 100644
--- a/.kitchen.yml
+++ b/.kitchen.yml
@@ -9,18 +9,16 @@ platforms:
   - name: ubuntu-14.04
     driver_config:
       box: 'ubuntu/trusty64'
-      box_url: 'http://files.vagrantup.com/trusty64.box'
       customize:
         memory: 2048
     attributes:
       zookeeper:
-        service_style: "upstart"
+        service_style: "runit"
       apache_kafka:
-        service_style: "upstart"
+        service_style: "runit"
   - name: debian-7.8
     driver_config:
       box: 'chef/debian-7.8'
-      box_url: 'https://atlas.hashicorp.com/chef/boxes/debian-7.8'
       customize:
         memory: 2048
     run_list:
@@ -29,7 +27,7 @@ platforms:
       zookeeper:
         service_style: "runit"
       apache_kafka:
-        service_style: "init.d"
+        service_style: "runit"

 suites:
   - name: default

```

(I also changed the zookeeper to runit for ubuntu, since it's already in use for debian)

I did have more than one successful run last week, but I haven't yet today, just sharing the changes I made to .kitchen.yml

**Comment by justizin** [comment]
Had to lock the runit cookbook to 1.6.0 (pushed a change to Berksfile in this branch), the version in my ~/.berkshelf.d from last week, I got past an issue where 1.7.2 failed entirely.  Expect this can be removed in time.

Had a successful run against just this branch.  Sending my kitchen.yml changes in another PR, since in general, the current .kitchen.yml just won't work on a fresh machine afaict - I have four that I run it on.

**Comment by mathyourlife** [comment]
Nice, that worked for me!

:+1:
